java/ql/src/semmle/code/java/dataflow/NullGuards.qll

```diff
@@ -44,21 +44,32 @@ Expr clearlyNotNullExpr(Expr reason) {
   or
   result instanceof ArrayCreationExpr and reason = result
   or
+  result instanceof FunctionalExpr and reason = result
+  or
   result instanceof TypeLiteral and reason = result
   or
   result instanceof ThisAccess and reason = result
   or
   result instanceof StringLiteral and reason = result
   or
+  // Add and AssignAdd performing String concatenation never have null result
   result instanceof AddExpr and result.getType() instanceof TypeString and reason = result
   or
+  result instanceof AssignAddExpr and result.getType() instanceof TypeString and reason = result
+  or
   exists(Field f |
     result = f.getAnAccess() and
-    (f.hasName("TRUE") or f.hasName("FALSE")) and
+    f.hasName(["TRUE", "FALSE"]) and
     f.getDeclaringType().hasQualifiedName("java.lang", "Boolean") and
     reason = result
   )
   or
+  result.getType() instanceof PrimitiveType and not result instanceof TypeAccess and reason = result
+  or
+  // Ignore RValue because that is covered by SsaVariable check below and correctly
+  // reports the assigned value as 'reason'
+  result instanceof CompileTimeConstantExpr and not result instanceof RValue and reason = result
```

Comment:
`null` is a compile-time constant expression.

Reply:
Also, are there any compile-time constant operations that aren't already included directly in this predicate? If so, should they? (and leading to making this disjunct superfluous?)

Reply:
> `null` is a compile-time constant expression.

At least according to the language specification, it is not. See [JLS 16 §15.29](https://docs.oracle.com/javase/specs/jls/se16/html/jls-15.html#jls-15.29):
> Literals of primitive type (§3.10.1, §3.10.2, §3.10.3, §3.10.4), string literals (§3.10.5), and text blocks (§3.10.6)

All of the other definitions rely on this point, so a `NullLiteral` cannot be a compile time constant.

Though you are right, with the check for primitive type below, the check for a `CompileTimeConstantExpr` here would probably be redundant.

Reply:
> At least according to the language specification, it is not. See [JLS 16 §15.29](https://docs.oracle.com/javase/specs/jls/se16/html/jls-15.html#jls-15.29):

Ah, right, I missed that (and the fact that `CompileTimeConstantExpr` indeed doesn't include `null`). But I'd expect all (or most of) the operations that make up compile-time constants to be included here, thus making it redundant.